doc/lenses.rst

```diff
@@ -3,16 +3,494 @@
+   std::optional<Digit> digit = view(firstDigit, mouse);
+    
+   auto firstLegPosition = attr(&Mouse::legs) // vector<Leg>
+           | first                            // optional<Leg>
+           | with_opt(legPosition);           // optional<int>
+   std::optional<int> position = view(firstLegPosition, mouse);
+
+This should be safe to use, but be weary of using it with models that
+have optionals as legitimate values. Using the less ambiguous
+``map_opt`` and ``bind_opt`` is preffered.
+
+Of course, we also provide a lens for falling back to either a
+default constructed value or a fallback value with ``value_or`` and
+``or_default``:
+
+.. code-block:: c++
+
+   auto firstLegPosition = attr(&Mouse::legs) // vector<Leg>
+           | first                            // optional<Leg>
+           | map_opt(legPosition);            // optional<int>
+    
+   auto withDefault = firstLegPosition | or_default; // default constructed
+   // auto withDefault = firstLegPosition | value_or(); // equivalent
+   auto withFallback = firstLegPosition | value_or(-1); // fallback to -1
+ 
+   int position = view(withFallback, mouse);
+
+
+.. _maybe monad: https://en.wikipedia.org/wiki/Monad_(functional_programming)
+
+.. _dynamic-lenses:
+
+Dynamic Lenses
+--------------
+
+You've probably noticed that all of our lenses have the type ``auto``
+in the previous examples. This is because VLLs rely on *compile-time
+type information* to implement ``view``, ``set`` and ``over``, and the
+resulting types are somewhat cryptic... This is fine for composing
+lenses at compile time, but here's the catch:
+
+.. code-block:: c++
+
+   struct Tail {
+       int position;
+       optional<box<Tail>> tail;
+   };
+    
+   auto tail = attr(&Tail::tail) | value_or() | unbox;
+   auto position = attr(&Tail::position);
+    
+   auto lens1 = tail | position;        // lens<Tail, int>
+   auto lens2 = tail | tail | position; // lens<Tail, int>
+    
+   static_assert(std::is_same_v<decltype(lens1), decltype(lens2)>,
+                 "Not the same types!");
+
+This means that you can't have this kind of pattern:
+
+.. code-block:: c++
+
+   auto tailPositionAt(int index) {
+       auto resultLens = position;
+       while(index-- > 0) {
+           resultLens = tail | resultLens; // won't compile, the type changed!
+       }
+       return resultLens;
+   }
+
+We need a way to store ``lens1`` and ``lens2`` in the same type,
+because they satisfy the same interface that we defined earlier (they
+are both, conceptually, ``lens<Tail, int>``).
+
+This is where *type erasure* comes in:
+
+.. code-block:: c++
+
+   #include <lager/lens.hpp> // type erased lenses
+    
+   lens<Tail, int> tailPositionAt(int index) {
+       lens<Tail, int> resultLens = position;
+       while (index-- > 0) {
+           resultLens = tail | resultLens; // this works now
+       }
+       return resultLens;
+   }
+
+``<lager/lens.hpp>`` provides a type erased lens for this very
```

Comment:
Starting a sentence with a file looks a bit weird, I'd suggest `The lager/lens.hpp header provides...` for something more gramatical.